File: src/main/java/net/coderextreme/data/HAnimModelsHandsFeet.java
package net.coderextreme.data;
import org.web3d.x3d.jsail.*;
import org.web3d.x3d.jsail.CADGeometry.*;
import org.web3d.x3d.jsail.Core.*;
import org.web3d.x3d.jsail.CubeMapTexturing.*;
import org.web3d.x3d.jsail.DIS.*;
import org.web3d.x3d.jsail.EnvironmentalEffects.*;
import org.web3d.x3d.jsail.EnvironmentalSensor.*;
import org.web3d.x3d.jsail.EventUtilities.*;
import org.web3d.x3d.jsail.Followers.*;
import org.web3d.x3d.jsail.Geometry2D.*;
import org.web3d.x3d.jsail.Geometry3D.*;
import org.web3d.x3d.jsail.Geospatial.*;
import org.web3d.x3d.jsail.Grouping.*;
import org.web3d.x3d.jsail.HAnim.*;
import org.web3d.x3d.jsail.Interpolation.*;
import org.web3d.x3d.jsail.KeyDeviceSensor.*;
import org.web3d.x3d.jsail.Layering.*;
import org.web3d.x3d.jsail.Layout.*;
import org.web3d.x3d.jsail.Lighting.*;
import org.web3d.x3d.jsail.NURBS.*;
import org.web3d.x3d.jsail.Navigation.*;
import org.web3d.x3d.jsail.Networking.*;
import org.web3d.x3d.jsail.ParticleSystems.*;
import org.web3d.x3d.jsail.Picking.*;
import org.web3d.x3d.jsail.PointingDeviceSensor.*;
import org.web3d.x3d.jsail.Rendering.*;
import org.web3d.x3d.jsail.RigidBodyPhysics.*;
import org.web3d.x3d.jsail.Scripting.*;
import org.web3d.x3d.jsail.Shaders.*;
import org.web3d.x3d.jsail.Shape.*;
import org.web3d.x3d.jsail.Sound.*;
import org.web3d.x3d.jsail.Text.*;
import org.web3d.x3d.jsail.Texturing3D.*;
import org.web3d.x3d.jsail.Texturing.*;
import org.web3d.x3d.jsail.Time.*;
import org.web3d.x3d.jsail.VolumeRendering.*;
import org.web3d.x3d.jsail.fields.*;
import java.util.ArrayList;
import java.util.List;
import net.coderextreme.X3DRoots;
public class HAnimModelsHandsFeet implements X3DRoots {
  public static void main(String[] args) {
    ConfigurationProperties.setXsltEngine(ConfigurationProperties.XSLT_ENGINE_NATIVE_JAVA);
    ConfigurationProperties.setDeleteIntermediateFiles(false);
    ConfigurationProperties.setStripTrailingZeroes(true);
    ConfigurationProperties.setStripDefaultAttributes(true);
    X3D model = new HAnimModelsHandsFeet().getRootNodeList().get(0); // only get one root node
    System.out.print(model.validationReport().trim());
    model.toFileX3D("../data/HAnimModelsHandsFeet.new.java.x3d");
    model.toFileJSON("../data/HAnimModelsHandsFeet.new.java.x3dj");
    }
    public List<X3D> getRootNodeList() {
    	List<X3D> list = new ArrayList<X3D>(1);
    	list.add(initialize());
    	return list;
    }
    public X3D initialize() {
      X3D X3D0 =  new X3D().setProfile("Immersive").setVersion("4.0")
      .setHead(new head()
        .addComponent(new component().setName("HAnim").setLevel(1))
        .addMeta(new meta().setName("title").setContent("HAnimModelsHandsFeet.x3d"))
        .addMeta(new meta().setName("description").setContent("Left and right hands and feet, using high-fidelity definitions for HAnim version 2.0"))
        .addMeta(new meta().setName("creator").setContent("Kwan Hee YOO and Don Brutzman"))
        .addMeta(new meta().setName("created").setContent("8 February 2015"))
        .addMeta(new meta().setName("modified").setContent("Mon, 15 Sep 2025 05:20:09 GMT"))
        .addMeta(new meta().setName("warning").setContent("not yet to scale"))
        .addMeta(new meta().setName("warning").setContent("TODO will X3D HAnim component add a new level to support LOA-4 functionality?"))
        .addMeta(new meta().setName("TODO").setContent("how to have HAnimHumanoid root with Inline IMPORT/EXPORT of limbs?"))
        .addMeta(new meta().setName("Image").setContent("HAnimModelsHandsFeet.png"))
        .addMeta(new meta().setName("Image").setContent("HAnimModelsHandsFeetWithFour1mGrids.png"))
        .addMeta(new meta().setName("reference").setContent("https://www.web3d.org/working-groups/humanoid-animation-HAnim"))
        .addMeta(new meta().setName("reference").setContent("https://www.web3d.org/documents/specifications/19774/V2.0"))
        .addMeta(new meta().setName("reference").setContent("https://www.web3d.org/specifications/X3Dv4/ISO-IEC19775-1v4-IS/Part01/components/hanim.html"))
        .addMeta(new meta().setName("subject").setContent("X3D HAnim humanoid animation"))
        .addMeta(new meta().setName("identifier").setContent("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelsHandsFeet.x3d")))
      .setScene(new Scene()
        .addChild(new WorldInfo().setTitle("HAnimModelsHandsFeet.x3d"))
        .addChild(new Viewpoint().setDescription("Hands and feet 10m"))
        .addChild(new Viewpoint().setDescription("Hands and feet 1.7m").setPosition(new float[] {0f ,0f ,1.7f }))
        .addChild(new Transform().setTranslation(new float[] {-1f ,1f ,0f })
          .addChild(new Inline().setUrl(new MFString0().getArray()))
          .addChild(new Transform().setDEF("GridXY_20x20Fixed_AdjustScale").setScale(new float[] {0.1f ,0.1f ,0.1f })
            .addChild(new Inline().setDEF("GridXY_20x20Fixed").setUrl(new MFString1().getArray()))))
        .addChild(new Transform().setTranslation(new float[] {1f ,1f ,0f })
          .addChild(new Inline().setUrl(new MFString2().getArray()))
          .addChild(new Transform().setUSE("GridXY_20x20Fixed_AdjustScale")))
        .addChild(new Transform().setTranslation(new float[] {-1f ,-1f ,0f })
          .addChild(new Inline().setUrl(new MFString3().getArray()))
          .addChild(new Transform().setUSE("GridXY_20x20Fixed_AdjustScale")))
        .addChild(new Transform().setTranslation(new float[] {1f ,-1f ,0f })
          .addChild(new Inline().setUrl(new MFString4().getArray()))
          .addChild(new Transform().setUSE("GridXY_20x20Fixed_AdjustScale"))));
    return X3D0;
    }
private class MFString0 {
  private org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"HAnimModelHandLeft.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandLeft.x3d","HAnimModelHandLeft.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandLeft.wrl"});
  }
}
private class MFString1 {
  private org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"../../Savage/Tools/Authoring/GridXY_20x20Fixed.x3d","https://savage.nps.edu/Savage/Tools/Authoring/GridXY_20x20Fixed.x3d","../../Savage/Tools/Authoring/GridXY_20x20Fixed.wrl","https://savage.nps.edu/Savage/Tools/Authoring/GridXY_20x20Fixed.wrl"});
  }
}
private class MFString2 {
  private org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"HAnimModelHandRight.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandRight.x3d","HAnimModelHandRight.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandRight.wrl"});
  }
}
private class MFString3 {
  private org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"HAnimModelFootLeft.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootLeft.x3d","HAnimModelFootLeft.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootLeft.wrl"});
  }
}
private class MFString4 {
  private org.web3d.x3d.jsail.fields.MFString getArray() {
    return new org.web3d.x3d.jsail.fields.MFString(new java.lang.String[] {"HAnimModelFootRight.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootRight.x3d","HAnimModelFootRight.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootRight.wrl"});
  }
}
}

File: src/main/node/net/coderextreme/data/HAnimModelsHandsFeet.js
import java from 'node-java';
import util from 'util';
import autoclass from '../../../X3Dautoclass.js';
var ConfigurationProperties = autoclass.ConfigurationProperties;
ConfigurationProperties.showDefaultAttributes = false;
ConfigurationProperties.xsltEngine = ConfigurationProperties.XSLT_ENGINE_NATIVE_JAVA;
ConfigurationProperties.deleteIntermediateFiles = false;
ConfigurationProperties.setStripTrailingZeroes(true);
ConfigurationProperties.setStripDefaultAttributes(true);
      var X3D0 =  (new autoclass.X3D()).setProfile("Immersive").setVersion("4.0")
      .setHead(new autoclass.head()
        .addComponent(new autoclass.component().setName("HAnim").setLevel(1))
        .addMeta(new autoclass.meta().setName("title").setContent("HAnimModelsHandsFeet.x3d"))
        .addMeta(new autoclass.meta().setName("description").setContent("Left and right hands and feet, using high-fidelity definitions for HAnim version 2.0"))
        .addMeta(new autoclass.meta().setName("creator").setContent("Kwan Hee YOO and Don Brutzman"))
        .addMeta(new autoclass.meta().setName("created").setContent("8 February 2015"))
        .addMeta(new autoclass.meta().setName("modified").setContent("Mon, 15 Sep 2025 05:20:09 GMT"))
        .addMeta(new autoclass.meta().setName("warning").setContent("not yet to scale"))
        .addMeta(new autoclass.meta().setName("warning").setContent("TODO will X3D HAnim component add a new level to support LOA-4 functionality?"))
        .addMeta(new autoclass.meta().setName("TODO").setContent("how to have HAnimHumanoid root with Inline IMPORT/EXPORT of limbs?"))
        .addMeta(new autoclass.meta().setName("Image").setContent("HAnimModelsHandsFeet.png"))
        .addMeta(new autoclass.meta().setName("Image").setContent("HAnimModelsHandsFeetWithFour1mGrids.png"))
        .addMeta(new autoclass.meta().setName("reference").setContent("https://www.web3d.org/working-groups/humanoid-animation-HAnim"))
        .addMeta(new autoclass.meta().setName("reference").setContent("https://www.web3d.org/documents/specifications/19774/V2.0"))
        .addMeta(new autoclass.meta().setName("reference").setContent("https://www.web3d.org/specifications/X3Dv4/ISO-IEC19775-1v4-IS/Part01/components/hanim.html"))
        .addMeta(new autoclass.meta().setName("subject").setContent("X3D HAnim humanoid animation"))
        .addMeta(new autoclass.meta().setName("identifier").setContent("https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelsHandsFeet.x3d")))
      .setScene(new autoclass.Scene()
        .addChild(new autoclass.WorldInfo().setTitle("HAnimModelsHandsFeet.x3d"))
        .addChild(new autoclass.Viewpoint().setDescription("Hands and feet 10m"))
        .addChild(new autoclass.Viewpoint().setDescription("Hands and feet 1.7m").setPosition(java.newArray("float", [java.newFloat(0), java.newFloat(0), java.newFloat(1.7)])))
        .addChild(new autoclass.Transform().setTranslation(java.newArray("float", [java.newFloat(-1), java.newFloat(1), java.newFloat(0)]))
          .addChild(new autoclass.Inline().setUrl(java.newArray("java.lang.String", ["HAnimModelHandLeft.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandLeft.x3d","HAnimModelHandLeft.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandLeft.wrl"])))
          .addChild(new autoclass.Transform().setDEF("GridXY_20x20Fixed_AdjustScale").setScale(java.newArray("float", [java.newFloat(0.1), java.newFloat(0.1), java.newFloat(0.1)]))
            .addChild(new autoclass.Inline().setDEF("GridXY_20x20Fixed").setUrl(java.newArray("java.lang.String", ["../../Savage/Tools/Authoring/GridXY_20x20Fixed.x3d","https://savage.nps.edu/Savage/Tools/Authoring/GridXY_20x20Fixed.x3d","../../Savage/Tools/Authoring/GridXY_20x20Fixed.wrl","https://savage.nps.edu/Savage/Tools/Authoring/GridXY_20x20Fixed.wrl"])))))
        .addChild(new autoclass.Transform().setTranslation(java.newArray("float", [java.newFloat(1), java.newFloat(1), java.newFloat(0)]))
          .addChild(new autoclass.Inline().setUrl(java.newArray("java.lang.String", ["HAnimModelHandRight.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandRight.x3d","HAnimModelHandRight.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelHandRight.wrl"])))
          .addChild(new autoclass.Transform().setUSE("GridXY_20x20Fixed_AdjustScale")))
        .addChild(new autoclass.Transform().setTranslation(java.newArray("float", [java.newFloat(-1), java.newFloat(-1), java.newFloat(0)]))
          .addChild(new autoclass.Inline().setUrl(java.newArray("java.lang.String", ["HAnimModelFootLeft.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootLeft.x3d","HAnimModelFootLeft.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootLeft.wrl"])))
          .addChild(new autoclass.Transform().setUSE("GridXY_20x20Fixed_AdjustScale")))
        .addChild(new autoclass.Transform().setTranslation(java.newArray("float", [java.newFloat(1), java.newFloat(-1), java.newFloat(0)]))
          .addChild(new autoclass.Inline().setUrl(java.newArray("java.lang.String", ["HAnimModelFootRight.x3d","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootRight.x3d","HAnimModelFootRight.wrl","https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Characters/HAnimModelFootRight.wrl"])))
          .addChild(new autoclass.Transform().setUSE("GridXY_20x20Fixed_AdjustScale"))))      ;
    X3D0.toFileX3D("../data/HAnimModelsHandsFeet.new.node.x3d");
    X3D0.toFileJSON("../data/HAnimModelsHandsFeet.new.node.x3dj");
    process.exit(0);
